Websocket and Ingame API changes

FFFFFFFXXXXXXX opened the pull request:
I put together a quick first version.
Is this roughly what you had in mind?

Im unsure how to pass the payload-data to the user. Currently I just pass the raw json string through, without any parsing or validation.

FelixArbeithuber committed:
rough first draft

FelixArbeithuber committed:
added errors and raw data pass-through

FelixArbeithuber committed:
oops

Leastrio commented:
If i had to be honest, i haven't used rust in a good few months so I am a bit rusty at it. But at first glance this looks pretty good. I dont remember exactly how WS events look but I was thinking of making a python/rust script to auto generate some structures for everything depending on the schema. We could possibly have the events deserialize into those structs.

FFFFFFFXXXXXXX commented:
I played around with the openapi schemas a little but I dont think auto-generation will work.
The "official" openapi.json you can get from the api /swagger/v3/openapi.json is incomplete and all the other schemas I found were incorrect.

For example the "end-of-game" api section is missing from the "official" schema.

And other schemas have incorrect versions of the "end-of-game" api section.
See https://www.mingweisamuel.com/lcu-schema/tool/#/Plugin%20lol-end-of-game/GetLolEndOfGameV1EogStatsBlock 
vs
[test_eog-stats-block](https://github.com/Leastrio/Shaco/files/10787030/eog-stats-block.txt)

FelixArbeithuber committed:
implement streams for ingame.rs and ws.rs

Leastrio commented:
well thats fun.. We could look to see how other libraries handle this and figure out from there.

FelixArbeithuber committed:
Merge remote-tracking branch 'upstream/main'

# Conflicts:
#	src/utils/process_info.rs

FFFFFFFXXXXXXX commented:
Instead of the WSClient::next_event() function I implemented the futures_util::Stream trait for the websocketclient, which provides the same functionality + some more.

I also sneaked in an EventStream struct for the ingame_api which polls the /liveclientdata api for you. If you prefer I can put it in a seperate pull-request.

From what I can find (via a quick google search) most lcu websocket libraries just return parsed json.
In python the obvious way is to provide a dictionary of the data e.g. https://lcu-driver.readthedocs.io/en/latest/quickstart.html#websocket.
In JS its the same but with a JSON object https://www.npmjs.com/package/league-connect
I found one C# library https://github.com/Ponita0/PoniLCU which also just returns a general JSON type.

So it seems like we will need to do something similar.
If we return a serde_json::Value (from serde_json) the user will have to use the serde_json crate, but if we dont and return a plain String the user will have to parse it anyways. So im currently leaning towards returning a serde_json::Value

What do you think?

Leastrio commented:
yeah you can include it in this pr if you'd like. My last idea was using serde Value so ig we can use it.

Leastrio commented:
also, what are your thoughts on switching from reqwest to hyper? seems to be a lot more lighter

FFFFFFFXXXXXXX commented:
Alright, I'll probably have the final commit for this PR ready tomorrow.

I did a little test of reqwest vs hyper polling /livegamedata/allgamedata and /livegamedata/eventdata each for about 60s every 500ms with:

```rust
use std::time::Duration;
use futures_util::TryFutureExt;

use tokio::time::Instant;

#[tokio::test]
async fn hyper() {
    let tls_connector = {
        let cert =
            native_tls::Certificate::from_pem(include_bytes!("../src/riotgames.pem")).unwrap();
        native_tls::TlsConnector::builder()
            .add_root_certificate(cert)
            .build()
            .unwrap()
    };

    let tokio_tls_connector = tokio_native_tls::TlsConnector::from(tls_connector);
    let mut hyper_http_connector = hyper::client::HttpConnector::new();
    hyper_http_connector.enforce_http(false);
    let connector = hyper_tls::HttpsConnector::from((hyper_http_connector, tokio_tls_connector));

    let client: hyper::client::Client<_, hyper::body::Body> = hyper::client::Client::builder()
        .pool_idle_timeout(None)
        .build(connector);

    let mut timer = tokio::time::interval(Duration::from_millis(500));
    loop {
        timer.tick().await;
        let now = Instant::now();
        let res = client
            .get(hyper::Uri::from_static(
                "https://127.0.0.1:2999/liveclientdata/eventdata",
            ))
            .await
            .unwrap();
        let status = res.status();
        let data = hyper::body::to_bytes(res).await;
        println!("{}: {}ms", status, now.elapsed().as_millis());
    }
}

#[tokio::test]
async fn reqwest() {
    let cert = reqwest::Certificate::from_pem(include_bytes!("../src/riotgames.pem")).unwrap();
    let mut headers = reqwest::header::HeaderMap::new();

    let client = reqwest::ClientBuilder::new()
        .add_root_certificate(cert)
        .default_headers(headers)
        .pool_idle_timeout(None)
        .build()
        .unwrap();

    let mut timer = tokio::time::interval(Duration::from_millis(500));
    loop {
        timer.tick().await;
        let now = Instant::now();
        let res = client
            .get("https://127.0.0.1:2999/liveclientdata/eventdata")
            .send()
            .await
            .unwrap();
        let status = res.status();
        let data = res.text().await.unwrap();
        println!("{}: {}ms", status, now.elapsed().as_millis());
    }
}
```

Maybe I missed some optimisation, but there doesnt seem to be a big difference (at least in terms of latency)

[reqwest_vs_hyper1.xlsx](https://github.com/Leastrio/Shaco/files/10807956/reqwest_vs_hyper1.xlsx)
[reqwest_vs_hyper2.xlsx](https://github.com/Leastrio/Shaco/files/10807957/reqwest_vs_hyper2.xlsx)

I havent checked out the size of the compiled binaries. Maybe its worth it for that reason

Leastrio commented:
hmmmm.. I just read what sylv said in the riot dev's server and maybe ill rethink switching to hyper. Im going to practice a little bit of rust in the next few days since I have used it in a while and after fully review your changes!

FelixArbeithuber committed:
final commit for PR, change Websocket Event deserialization to serde_json::Value

FelixArbeithuber committed:
checking for wrong Error type

AlsoSylv commented on src/utils/request.rs:
```diff
@@ -15,5 +17,6 @@ pub fn build_reqwest_client(
     Ok(reqwest::ClientBuilder::new()
         .add_root_certificate(cert)
         .default_headers(headers)
+        .pool_idle_timeout(Some(Duration::from_secs(5 * 60)))
```
This is here on purpose?

AlsoSylv replied:
This seems to be for event stream in the in_game.rs module, but it doesn't actually use the polling rate set there?

FFFFFFFXXXXXXX replied:
I initially added for my reqwest vs hyper benchmarks to make sure the tcp connection doesnt get restarted during the test. But the default of 90s should be fine so ill remove it

AlsoSylv commented on src/utils/process_info.rs:
```diff
@@ -10,45 +8,31 @@ const TARGET_PROCESS: &str = "LeagueClientUx.";
-}
+        .find(|p| p.name() == TARGET_PROCESS)
+        .map(|p| p.cmd())
+        .ok_or("no LoL process")?;
```
If the error messages are going to be changed, it makes sense to use an error enum instead of a new &str

AlsoSylv commented on src/utils/process_info.rs:
```diff
@@ -10,45 +8,31 @@ const TARGET_PROCESS: &str = "LeagueClientUx.";
+        .find(|arg| arg.starts_with("--app-port="))
+        .map(|arg| arg[11..].parse::<u16>().ok())
+        .flatten()
+        .ok_or("no app_port")?;
```
Same here

AlsoSylv commented on src/utils/process_info.rs:
```diff
@@ -10,45 +8,31 @@ const TARGET_PROCESS: &str = "LeagueClientUx.";
+        .into_iter()
+        .find(|arg| arg.starts_with("--remoting-auth-token="))
+        .map(|arg| arg[22..].to_string())
+        .ok_or("no remoting_auth_token")?;
```
And here

AlsoSylv commented on src/utils/process_info.rs:
```diff
@@ -10,45 +8,31 @@ const TARGET_PROCESS: &str = "LeagueClientUx.";
+    let port = args
+        .into_iter()
+        .find(|arg| arg.starts_with("--app-port="))
+        .map(|arg| arg[11..].parse::<u16>().ok())
```
Is there a reason this is parsed to an u16?

FFFFFFFXXXXXXX replied:
port numbers fit exactly into an u16, if the number is outside of that range it is invalid

AlsoSylv replied:
Right, but it doesn't actually need to be stored as a u16, and will remove the parse step to store it as a string

AlsoSylv commented on src/utils/process_info.rs:
```diff
@@ -10,45 +8,31 @@ const TARGET_PROCESS: &str = "LeagueClientUx.";
+    let auth_token = args
+        .into_iter()
+        .find(|arg| arg.starts_with("--remoting-auth-token="))
+        .map(|arg| arg[22..].to_string())
```
Shouldn't this use strip_prefix?

AlsoSylv commented on src/ws.rs:
```diff
@@ -1,78 +1,86 @@
-        let args = extract_info(process)?;
-        let auth_token = encode_token(&args.0);
+    pub async fn connect() -> Result<Self, LcuWebsocketError> {
+        let (auth_token, port) = get_auth_info().map_err(|_| LcuWebsocketError::LcuNotAvailable)?;
```
Does `.unwrap_or(Err(LcuWebsocketError::LcuNotAvailable)?;` work? It would be better for readability

AlsoSylv commented on src/ws.rs:
```diff
@@ -1,78 +1,86 @@
+        let (auth_token, port) = get_auth_info().map_err(|_| LcuWebsocketError::LcuNotAvailable)?;
 
-        let cert = native_tls::Certificate::from_pem(include_bytes!("./riotgames.pem"))?;
+        let cert = native_tls::Certificate::from_pem(include_bytes!("./riotgames.pem"))
```
It should be possible to safely unwrap this

FFFFFFFXXXXXXX replied:
Does it matter if we expect instead of unwrap? The "message" just makes it easier to find in case the cert is actually invalid at some point

AlsoSylv replied:
The cert itself should always be valid at this step, it's when making requests that it would be invalid

AlsoSylv commented on src/ws.rs:
```diff
@@ -1,78 +1,86 @@
-            );
-        }
+            .build()
+            .map_err(|_| LcuWebsocketError::AuthError)?;
```
`.unwrap_or()` could be used here

AlsoSylv commented on src/ws.rs:
```diff
@@ -1,78 +1,86 @@
+        url.headers_mut().insert(
+            "Authorization",
+            HeaderValue::from_str(format!("Basic {auth_token}").as_str())
+                .map_err(|_| LcuWebsocketError::AuthError)?,
```
This should also be `.unwrap_or()`

AlsoSylv commented on src/ws.rs:
```diff
@@ -1,78 +1,86 @@
-            tokio_tungstenite::connect_async_tls_with_config(url, None, Some(connector)).await?;
+        let mut url = format!("wss://127.0.0.1:{port}")
+            .into_client_request()
+            .map_err(|_| LcuWebsocketError::AuthError)?;
```
Here too

AlsoSylv commented on src/ws.rs:
```diff
@@ -1,78 +1,86 @@
+        let (ws_stream, _response) =
+            tokio_tungstenite::connect_async_tls_with_config(url, None, Some(connector))
+                .await
+                .map_err(|_| LcuWebsocketError::Disconnected)?;
```
And here

AlsoSylv commented on src/ws.rs:
```diff
@@ -1,78 +1,86 @@
-        self.write
-            .send(Message::text(format!("[5, {}]", event)))
-            .await?;
+    pub async fn subscribe(&mut self, event: LcuSubscriptionType) -> Result<(), LcuWebsocketError> {
```
The code here is duplicated, and could be templated into something like
```rs
    pub async fn event(
        &mut self,
        event: LcuSubscriptionType,
        event_id: i8,
    ) -> Result<(), LcuWebsocketError> {
        self.0
            .send(Message::text(format!("[{event_id}, \"{event}\"]")))
            .await
            .map_err(|e| match e {
                Error::ConnectionClosed | Error::AlreadyClosed => LcuWebsocketError::Disconnected,
                _ => LcuWebsocketError::SendError,
            })
    }

FFFFFFFXXXXXXX replied:
should we really provide an API for a variable event_id? As far as i know the only useful ids are 5/6 for subscribe/unsubscribe?

unless you mean something like this?

```rust
pub async fn subscribe(&mut self, subscription: LcuSubscriptionType) -> Result<(), LcuWebsocketError> {
        self.send_message(5, subscription).await
    }

pub async fn unsubscribe(
    &mut self,
    subscription: LcuSubscriptionType,
) -> Result<(), LcuWebsocketError> {
    self.send_message(6, subscription).await
}

async fn send_message(
    &mut self,
    event_id: u8,
    subscription: LcuSubscriptionType,
) -> Result<(), LcuWebsocketError> {
    self.0
        .send(Message::text(format!("[{event_id}, \"{subscription}\"]")))
        .await
        .map_err(|e| match e {
            Error::ConnectionClosed | Error::AlreadyClosed => {
                LcuWebsocketError::Disconnected(e.to_string())
            }
            _ => LcuWebsocketError::SendError,
        })
}
```

AlsoSylv replied:
Yes that is exactly what I mean

AlsoSylv commented on src/model/ws.rs:
```diff
@@ -0,0 +1,108 @@
+use std::error::Error;
```
This module needs doc comments

AlsoSylv commented on src/ingame.rs:
```diff
@@ -257,4 +259,96 @@ impl InGameClient {
+    pub async fn active_game(&self) -> bool {
+        let req = self
+            .reqwest_client
+            .head(format!("https://127.0.0.1:{}/Help", self.port)) // only /Help doesn't 404 during loading screen
```
This comment should be a doc comment like
```rs
/// sends a /Help request, as it is the only request that will not 404 during the loading screen

FelixArbeithuber committed:
hopefully fix all PR review comments

FelixArbeithuber committed:
cleanup imports

FelixArbeithuber committed:
oopsie (undo .unwrap_or())

FelixArbeithuber committed:
since the scope of this PR has already expanded massively I thought I would also take a look at the ingame API type definitions and add custom errors for the ingame API

FelixArbeithuber committed:
Fix turrets and inhibs + add comments

FelixArbeithuber committed:
minor naming change

FFFFFFFXXXXXXX commented:
I was a little bored on the weekend so I also added more exact types for the ingame API so the users of the library dont have to do as much "parsing" / "interpreting" or whatever you want to call it of the deserialized data

I also added some more tests.

FelixArbeithuber committed:
add Display trait to enums and add Fountains to Turret enum

Leastrio commented:
Appreciate all the commits! I will probably add some CI to verify it all and possibly merge this weekend(?)

Leastrio commented:
okay i cant figure out how to trigger CI on this pr so if you can do that it would be amazing

FFFFFFFXXXXXXX commented:
What do you mean by "trigger CI"?

Leastrio commented:
cause it to run on your pr. You can do that by making a commit, you could probably create an empty commit and push that.

FFFFFFFXXXXXXX committed:
Merge branch 'Leastrio:main' into main

Leastrio commented:
okay never mind ill fix the CI after i merge your pr

FelixArbeithuber committed:
fix rustfmt and clippy errors

FelixArbeithuber committed:
Merge remote-tracking branch 'origin/main'

FFFFFFFXXXXXXX commented:
I tried to fix the rustfmt error and the clippy I saw in the CI results.
Maybe this helps :)

Leastrio commented on src/model/ingame.rs:
```diff
@@ -1,191 +1,1012 @@
+        #[derive(Debug, Clone, Serialize, Deserialize)]
+        #[serde(deny_unknown_fields)]
+        #[serde(rename_all = "camelCase")]
+        pub struct Holder {
```
Why is `Holder` public here? Is it needed?

Leastrio commented:
Other than that, Sylv got everything else I wouldve commented on. CI works now so after that last comment gets resolved I shall merge :)

Leastrio commented:
Some docs would probably be good too, but I say we worry about that later

FelixArbeithuber committed:
remove unnecessary pub && add Enum::Unknown so deserialization doesn't fail if riot changes things

FFFFFFFXXXXXXX commented:
Thx, that "pub" was a copy and paste error.

Leastrio commented:
Appreciate all of the work youve done!
